### PR TITLE
Document text index commands

### PR DESCRIPTION
**Pull request by Hunterness**
Covers part of [this CIP](https://docs.google.com/document/d/1o1xZFtuwg-VKwNpG7lgcIzsNpKS_m4h_8PsTG7RqsZo/edit?usp=sharing). The other parts will come when they get full support.

### CODE REVIEW

File: cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/DocumentingTestBase.scala [hunk 1]
```diff
@@ -572,7 +573,8 @@ abstract class DocumentingTestBase extends JUnitSuite with DocumentationHelper w
   protected def databaseConfig(): util.Map[Setting[_], Object] =
     Map[Setting[_], Object](
       OnlineBackupSettings.online_backup_listen_address -> new SocketAddress("127.0.0.1", 0),
-      OnlineBackupSettings.online_backup_enabled -> java.lang.Boolean.FALSE
+      OnlineBackupSettings.online_backup_enabled -> java.lang.Boolean.FALSE,
+      GraphDatabaseInternalSettings.text_indexes_enabled -> java.lang.Boolean.TRUE
```
Comment (Hunterness) [inline cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/DocumentingTestBase.scala]: Do we need to document this? (that the flag is needed for the indexes)

File: cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/SchemaIndexTest.scala [hunk 2]
```diff
@@ -192,6 +195,52 @@ class SchemaIndexTest extends DocumentingTestBase with QueryStatisticsTestSuppor
...
+      text = "Token lookup indexes allow setting the index provider using the `OPTIONS` clause. " +
+        "Note that the index is not immediately available, but is created in the background.",
+      prepare = _ => executePreparationQueries(List("DROP INDEX node_label_lookup_index IF EXISTS")),
+      queryText = s"CREATE LOOKUP INDEX node_label_lookup_index_2 FOR (n) ON EACH labels(n) OPTIONS {indexProvider: '${TokenIndexProvider.DESCRIPTOR.name()}'}",
+      optionalResultExplanation = "Currently only one provider is available for token lookup indexes.",
+      assertions = _ => assertTokenLookupIndexExists("node_label_lookup_index_2", isNodeIndex = true)
+    )
+  }
+
+  @Test def create_text_index() {
```
Comment (Hunterness) [inline cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/SchemaIndexTest.scala:217]: I don't have any tests trying to use the index (not for lookup either I noticed), those only exists for btree. Do we want that kind of test?

Also, I don't know which of the indexes planner team have time to pick up and use in plans, but do we need to mention that somewhere in the docs?

File: cypher/cypher-docs/src/docs/dev/ql/administration/indexes-for-search-performance/index.asciidoc [hunk 3]
```diff
@@ -332,7 +366,8 @@ Listing indexes also allows for `WHERE` and `YIELD` clauses to filter the return
 
 [NOTE]
 ====
-The deprecated built-in procedures for listing indexes, such as `db.indexes`, work as before and are not affected by the <<administration-security-administration-database-indexes, `SHOW INDEXES` privilege>>.
+While the command for listing indexes require <<administration-security-administration-database-indexes, the `SHOW INDEX` privilege>>,
+the deprecated built-in procedures for listing indexes, such as `db.indexes`, work as before and are not affected by the privilege.
```
Comment (Lojjs) [inline cypher/cypher-docs/src/docs/dev/ql/administration/indexes-for-search-performance/index.asciidoc]: This looks like an unrelated change, which I thought we did already do 🤔 , or was that in some other place/version?

File: cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/DocumentingTestBase.scala (hunk 1 above)
Comment (Lojjs) [reply cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/DocumentingTestBase.scala]: Good question, I think we should discuss that with kernel team

File: cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/SchemaIndexTest.scala [hunk 4]
```diff
@@ -192,6 +195,52 @@ class SchemaIndexTest extends DocumentingTestBase with QueryStatisticsTestSuppor
       optionalResultExplanation = "Note that it can only be created once and that the index name must be unique.",
       assertions = _ => assertTokenLookupIndexExists("rel_type_lookup_index", isNodeIndex = false)
     )
+    prepareAndTestQuery(
+      title = "Create a token lookup index specifying the index provider",
+      text = "Token lookup indexes allow setting the index provider using the `OPTIONS` clause. " +
+        "Note that the index is not immediately available, but is created in the background.",
+      prepare = _ => executePreparationQueries(List("DROP INDEX node_label_lookup_index IF EXISTS")),
+      queryText = s"CREATE LOOKUP INDEX node_label_lookup_index_2 FOR (n) ON EACH labels(n) OPTIONS {indexProvider: '${TokenIndexProvider.DESCRIPTOR.name()}'}",
+      optionalResultExplanation = "Currently only one provider is available for token lookup indexes.",
```
Comment (Lojjs) [inline cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/SchemaIndexTest.scala]: I think it would be nice to name the provider in this sentence, like you did for text indexes.

File: cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/SchemaIndexTest.scala (hunk 2 above)
Comment (Lojjs) [reply cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/SchemaIndexTest.scala:217]: Using lookup is something that the planner team should have added in 4.3 I think, perhaps check with them if it is an oversight or if it is added somewhere else, like in the Query tuning chapter. 

I talked a bit with @tobias-johansson about indexes this morning and it seems a bit unclear if we will have planning support or not. I guess it would be worth mentioning somewhere that the indexes can only be created and not used for now, but not sure where it fits. Maybe we need to talk a bit about it with the people in the index channel and PM to give the correct message?

File: cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/ConstraintsTest.scala [hunk 5]
```diff
@@ -213,7 +213,7 @@ class ConstraintsTest extends DocumentingTestBase with SoftReset {
       text = "If it is unknown if a constraint exists or not but we want to make sure it does, we add the `IF NOT EXISTS`. " +
         "The node property existence constraint ensures that all nodes with a certain label have a certain property.",
       prepare = _ => executePreparationQueries(List("CREATE CONSTRAINT constraint_name IF NOT EXISTS FOR (book:Book) REQUIRE book.isbn IS UNIQUE")),
-      queryText = "CREATE CONSTRAINT constraint_name IF NOT EXISTS FOR (book:Book) REQUIRE book.isbn IS NOT NULL",
+      queryText = "CREATE CONSTRAINT constraint_name IF NOT EXISTS FOR (book:Book) REQUIRE book.isbn IS NOT NULL OPTIONS {}",
```
Comment (Lojjs) [inline cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/ConstraintsTest.scala]: Even if it is possible to add `OPTIONS {}` here for completeness, I would not include it in the example. Since it isn't required for doing IF NOT EXISTS but just redundant syntax to type, I think it is not helpful for users.

File: cypher/cypher-docs/src/docs/dev/ql/administration/indexes-for-search-performance/index.asciidoc [hunk 6]
```diff
@@ -110,37 +110,61 @@ Index provider and configuration can be specified using the `OPTIONS` clause.
 CREATE LOOKUP INDEX [index_name] [IF NOT EXISTS]
 FOR (n)
 ON EACH labels(n)
+[OPTIONS "{" option: value[, ...] "}"]
```
Comment (Lojjs) [inline cypher/cypher-docs/src/docs/dev/ql/administration/indexes-for-search-performance/index.asciidoc]: Strictly speaking , the `option: value[, ...]` isn't quite correct since there only exists one specific option for lookup and text indexes. Not sure if we want to reflect that here and in the glossary though, or leave it as is and just look at it as a general OPTIONS meta syntax?

File: cypher/cypher-docs/src/docs/dev/ql/administration/indexes-for-search-performance/index.asciidoc [hunk 7]
```diff
@@ -110,37 +110,61 @@ Index provider and configuration can be specified using the `OPTIONS` clause.
...
 ----
-| Drop an index, either a b-tree, fulltext, or token lookup index.
+| Drop an index of any index type.
 | The command is optionally idempotent, with the default behavior to throw an error if you attempt to drop the same index twice.
 With `IF EXISTS`, no error is thrown and nothing happens should the index not exist.
 
 | [source, cypher, role=noplay]
 ----
-SHOW [ALL\|BTREE\|FULLTEXT\|LOOKUP] INDEX[ES]
+SHOW [ALL\|BTREE\|FULLTEXT\|TEXT\|LOOKUP] INDEX[ES]
```
Comment (Lojjs) [inline cypher/cypher-docs/src/docs/dev/ql/administration/indexes-for-search-performance/index.asciidoc]: If want them alphabetically, LOOKUP should be before TEXT.

File: cypher/cypher-docs/src/docs/dev/ql/administration/indexes-for-search-performance/index.asciidoc [hunk 8]
```diff
@@ -288,7 +322,7 @@ Listing indexes can be done with `SHOW INDEXES`, which will produce a table with
 | `+`
 
 | `type`
-| The IndexType of this index (`BTREE`, `FULLTEXT`, or `LOOKUP`).
+| The IndexType of this index (`BTREE`, `FULLTEXT`, `TEXT`, or `LOOKUP`).
```
Comment (Lojjs) [inline cypher/cypher-docs/src/docs/dev/ql/administration/indexes-for-search-performance/index.asciidoc]: In most other places we mention `LOOKUP` before `TEXT`, which also happen to alphabetical order, maybe you want that order here too.

**Review by Lojjs — CHANGES_REQUESTED** [review]
The bigger question marks you have already pointed out yourself. Those need to be discussed. Other than that I just have minor things.

File: cypher/cypher-docs/src/docs/dev/ql/administration/indexes-for-search-performance/index.asciidoc (hunk 3 above)
Comment (Hunterness) [reply cypher/cypher-docs/src/docs/dev/ql/administration/indexes-for-search-performance/index.asciidoc]: It is higher up in the file, but I did add it here as well when I realized it might be good to have in both places

File: cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/ConstraintsTest.scala (hunk 5 above)
Comment (Hunterness) [reply cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/ConstraintsTest.scala]: I kind of wanted to have one example using it but not enough to add a whole new one but I agree with your point as well

File: cypher/cypher-docs/src/docs/dev/ql/administration/indexes-for-search-performance/index.asciidoc (hunk 6 above)
Comment (Hunterness) [reply cypher/cypher-docs/src/docs/dev/ql/administration/indexes-for-search-performance/index.asciidoc]: You can send in an empty config map as well :P It feels better to have the general syntax here and then specify what is allowed instead of restricting it here (also makes it easier to add other things to it :P)

File: cypher/cypher-docs/src/docs/dev/ql/administration/indexes-for-search-performance/index.asciidoc (hunk 7 above)
Comment (Hunterness) [reply cypher/cypher-docs/src/docs/dev/ql/administration/indexes-for-search-performance/index.asciidoc]: I'm planning on adding RANGE between BTREE and FULLTEXT... :P (since range is the common replacement for btree) and maybe text should be before fulltext as well though

Or I guess I could update to have all in alphabetical order

File: cypher/cypher-docs/src/docs/dev/ql/administration/indexes-for-search-performance/index.asciidoc (hunk 3 above)
Comment (Lojjs) [reply cypher/cypher-docs/src/docs/dev/ql/administration/indexes-for-search-performance/index.asciidoc]: Ok, just thought it looked familiar

File: cypher/cypher-docs/src/docs/dev/ql/administration/indexes-for-search-performance/index.asciidoc (hunk 6 above)
Comment (Lojjs) [reply cypher/cypher-docs/src/docs/dev/ql/administration/indexes-for-search-performance/index.asciidoc]: Ok, good point with the empty map. Let's keeps it as it is then

File: cypher/cypher-docs/src/docs/dev/ql/administration/indexes-for-search-performance/index.asciidoc (hunk 7 above)
Comment (Lojjs) [reply cypher/cypher-docs/src/docs/dev/ql/administration/indexes-for-search-performance/index.asciidoc]: I think either a logical or alphabetical order is ok, this is not really any of it :P

Comment (Hunterness) [reply cypher/cypher-docs/src/docs/dev/ql/administration/indexes-for-search-performance/index.asciidoc]: might just be easier to add it in alphabetical, then it takes less thinking to update :P

File: cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/DocumentingTestBase.scala (hunk 1 above)
Comment (Hunterness) [reply cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/DocumentingTestBase.scala]: Answer: No since it should go away before the release

File: cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/SchemaIndexTest.scala (hunk 2 above)
Comment (Hunterness) [reply cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/SchemaIndexTest.scala:217]: planner team have been notified about the missing lookup tests and that index usage tests are not done for text either

Answer: We don't document anything until supported by planner

File: cypher/cypher-docs/src/docs/dev/deprecations-additions-and-compatibility.asciidoc [hunk 9]
```diff
@@ -75,6 +75,45 @@ Replaced by:
...
+a|
+Token lookup indexes now allow an `OPTIONS` map to specify the index provider.
+
+
+a|
+label:functionality[]
+label:added[]
+[source, cypher, role="noheader"]
+----
+CREATE TEXT INDEX ...
```
Comment (Hunterness) [inline cypher/cypher-docs/src/docs/dev/deprecations-additions-and-compatibility.asciidoc:111]: update the btree deprecation part to show what part is replaced by text indexes once the [btree deprecation pr](https://github.com/neo4j/neo4j-documentation/pull/1256) is in

```
a|
label:functionality[]
label:deprecated[]
[source, cypher, role="noheader"]
----
CREATE BTREE INDEX ...
OPTIONS "{" indexProvider: 'lucene+native-3.0' "}"
----
a|
B-tree indexes are deprecated and will need migration before upgrading to 5.0.

B-tree indexes with the `lucene+native-3.0` index provider is replaced by:
[source, cypher, role="noheader"]
----
CREATE TEXT INDEX ...
----
```

**Comment by Hunterness** [comment]
[Document TEXT indexes](https://trello.com/c/2U55a2yK/2588-document-text-indexes)

**Review by Lojjs — APPROVED** [review]
Once we know that we will have planner support for text indexes, I think this is ready to go.

File: cypher/cypher-docs/src/docs/dev/ql/administration/indexes-for-search-performance/index.asciidoc [hunk 10]
```diff
@@ -127,22 +127,44 @@ ON [EACH] type(r)
...
 Index provider can be specified using the `OPTIONS` clause.
 
+| [source, cypher, role=noplay]
+----
+CREATE TEXT INDEX [index_name] [IF NOT EXISTS]
+FOR (n:LabelName)
+ON (n.propertyName)
+[OPTIONS "{" option: value[, ...] "}"]
+----
+| Create a text index on nodes.
```
Comment (Hunterness) [inline cypher/cypher-docs/src/docs/dev/ql/administration/indexes-for-search-performance/index.asciidoc]: Should we mention somewhere that text index only handles properties with string values? And if so, where would be good to add that? @Lojjs

Comment (Lojjs) [reply cypher/cypher-docs/src/docs/dev/ql/administration/indexes-for-search-performance/index.asciidoc]: That make sense yes, maybe we should extend the index for search performance introduction section `1. Indexes (types and limitations)` with that info?

File: cypher/refcard-tests/src/test/scala/org/neo4j/cypher/docgen/refcard/IndexTest.scala [hunk 11]
```diff
@@ -153,6 +153,22 @@ CREATE FULLTEXT INDEX rel_fulltext_index_name FOR ()-[r:HAS_PET|BROUGHT_PET]-()
...
+
+CREATE TEXT INDEX FOR (f:Friend) ON (f.email)
+###
+
+Create a text index on nodes with label `Friend` and property `email`.
+
+###assertion=create-index
+//
+
+CREATE TEXT INDEX text_index_name FOR ()-[h:HAS_PET]-() ON (h.since)
```
Comment (Lojjs) [inline cypher/refcard-tests/src/test/scala/org/neo4j/cypher/docgen/refcard/IndexTest.scala]: Related to you comment about text index only being on strings, just realised that `since` might not be the best example here since that is something you would typically model with a temporal or a number. Maybe change it to the name of the pet or something?

File: cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/SchemaIndexTest.scala [hunk 12]
```diff
@@ -206,6 +207,43 @@ class SchemaIndexTest extends DocumentingTestBase with QueryStatisticsTestSuppor
...
+        "Note that the index is not immediately available, but is created in the background.",
+      queryText = "CREATE TEXT INDEX node_index_name FOR (n:Person) ON (n.surname)",
+      optionalResultExplanation = "Note that text indexes do not support multiple properties and that the index name needs to be unique.",
+      assertions = _ => assertIndexWithNameExists("node_index_name", "Person", List("surname"))
+    )
+    testQuery(
+      title = "Create a relationship text index",
+      text = "A named text index on a single property for all relationships that have a particular relationship type can be created with `CREATE TEXT INDEX index_name FOR ()-[r:TYPE]-() ON (r.property)`. " +
+        "Note that the index is not immediately available, but is created in the background.",
+      queryText = "CREATE TEXT INDEX rel_index_name FOR ()-[r:KNOWS]-() ON (r.since)",
```
Comment (Lojjs) [inline cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/SchemaIndexTest.scala]: Related to you comment about text index only being on strings, just realised that since might not be the best example here since that is something you would typically model with a temporal or a number.

File: cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/SchemaIndexTest.scala [hunk 13]
```diff
@@ -206,6 +207,43 @@ class SchemaIndexTest extends DocumentingTestBase with QueryStatisticsTestSuppor
...
+      optionalResultExplanation = "Note that text indexes do not support multiple properties and that the index name needs to be unique.",
+      assertions = _ => assertIndexWithNameExists("rel_index_name", "KNOWS", List("since"))
+    )
+    testQuery(
+      title = "Create a text index only if it does not already exist",
+      text = "If it is unknown if an index exists or not but we want to make sure it does, we add `IF NOT EXISTS`.",
+      queryText = "CREATE TEXT INDEX node_index_name IF NOT EXISTS FOR (n:Person) ON (n.surname)",
+      optionalResultExplanation = "Note that the index will not be created if there already exists an index with the same name, same schema or both.",
```
Comment (Lojjs) [inline cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/SchemaIndexTest.scala]: Here we should also add something about the index type, since names should be unique across all index types, but schemas does not have to. This probably applies to all IF EXISTS tests, both older ones and text, range and point ones.

File: cypher/refcard-tests/src/test/scala/org/neo4j/cypher/docgen/refcard/IndexTest.scala (hunk 11 above)
Comment (Hunterness) [reply cypher/refcard-tests/src/test/scala/org/neo4j/cypher/docgen/refcard/IndexTest.scala]: I tried to find something that fitted on a relationship but it did not go well :P

**Review by Lojjs — APPROVED** [review]
Did not rebuild it again but looks like you fixed all the comments

File: cypher/cypher-docs/src/docs/dev/ql/administration/indexes-for-search-performance/index.asciidoc [hunk 14]
```diff
@@ -23,13 +23,19 @@ This comes at the cost of additional storage space and slower writes, so decidin
 Once an index has been created, it will be managed and kept up to date by the DBMS.
 Neo4j will automatically pick up and start using the index once it has been created and brought online.
 
-Cypher enables the creation of indexes on one or more properties for all nodes or relationships that have a given label or relationship type:
+There are multiple index types available, b-tree, fulltext, lookup, and text index types.
```
Comment (AlexicaWright) [inline cypher/cypher-docs/src/docs/dev/ql/administration/indexes-for-search-performance/index.asciidoc]:
```suggestion
There are multiple index types available: b-tree, fulltext, lookup, and text index types.
```

File: cypher/cypher-docs/src/docs/dev/ql/administration/indexes-for-search-performance/index.asciidoc [hunk 15]
```diff
@@ -23,13 +23,19 @@ This comes at the cost of additional storage space and slower writes, so decidin
...
+More information about fulltext indexes are found in <<administration-indexes-fulltext-search>>.
```
Comment (AlexicaWright) [inline cypher/cypher-docs/src/docs/dev/ql/administration/indexes-for-search-performance/index.asciidoc]:
```suggestion
See <<administration-indexes-fulltext-search>> for more information about fulltext indexes .
```

File: cypher/cypher-docs/src/docs/dev/ql/administration/indexes-for-search-performance/index.asciidoc [hunk 16]
```diff
@@ -23,13 +23,19 @@ This comes at the cost of additional storage space and slower writes, so decidin
...
+Token lookup indexes contains nodes with one or more label or relationship type, without any regard for properties.
```
Comment (AlexicaWright) [inline cypher/cypher-docs/src/docs/dev/ql/administration/indexes-for-search-performance/index.asciidoc]:
```suggestion
Token lookup indexes contain nodes with one or more labels or relationship types, without regard for any properties.
```

File: cypher/cypher-docs/src/docs/dev/ql/administration/indexes-for-search-performance/index.asciidoc [hunk 17]
```diff
@@ -23,13 +23,19 @@ This comes at the cost of additional storage space and slower writes, so decidin
...
+
+Cypher enables the creation of b-tree indexes on one or more properties for all nodes or relationships that have a given label or relationship type:
```
Comment (AlexicaWright) [inline cypher/cypher-docs/src/docs/dev/ql/administration/indexes-for-search-performance/index.asciidoc]:
```suggestion
Cypher enables the creation of b-tree indexes on one or more properties for all nodes or relationships with a given label or relationship type:
```

File: cypher/cypher-docs/src/docs/dev/ql/administration/indexes-for-search-performance/index.asciidoc [hunk 18]
```diff
@@ -23,13 +23,19 @@ This comes at the cost of additional storage space and slower writes, so decidin
...
 
 * An index that is created on a single property for any given label or relationship type is called a _single-property index_.
 * An index that is created on more than one property for any given label or relationship type is called a _composite index_.
 
 Differences in the usage patterns between composite and single-property indexes are described in <<administration-indexes-single-vs-composite-index>>.
 
+Text indexes are also a version of single-property indexes, with the limitation that they only knows about properties with string values.
```
Comment (AlexicaWright) [inline cypher/cypher-docs/src/docs/dev/ql/administration/indexes-for-search-performance/index.asciidoc]:
```suggestion
Additionally, a text index is a form of single-property index, with the limitation that it only recognizes properties with string values.
```

File: cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/SchemaIndexTest.scala [hunk 19]
```diff
@@ -206,6 +207,43 @@ class SchemaIndexTest extends DocumentingTestBase with QueryStatisticsTestSuppor
     )
   }
 
+  @Test def create_text_index() {
+    val textProvider = TextIndexProviderFactory.DESCRIPTOR.name()
+
+    testQuery(
+      title = "Create a node text index",
+      text = "A named text index on a single property for all nodes that have a particular label can be created with `CREATE TEXT INDEX index_name FOR (n:Label) ON (n.property)`. " +
```
Comment (AlexicaWright) [inline cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/SchemaIndexTest.scala]:
```suggestion
      text = "A named text index on a single property for all nodes with a particular label can be created with `CREATE TEXT INDEX index_name FOR (n:Label) ON (n.property)`. " +
```

File: cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/SchemaIndexTest.scala [hunk 20]
```diff
@@ -206,6 +207,43 @@ class SchemaIndexTest extends DocumentingTestBase with QueryStatisticsTestSuppor
...
+        "Note that the index is not immediately available, but is created in the background.",
+      queryText = "CREATE TEXT INDEX node_index_name FOR (n:Person) ON (n.surname)",
+      optionalResultExplanation = "Note that text indexes do not support multiple properties and that the index name needs to be unique.",
```
Comment (AlexicaWright) [inline cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/SchemaIndexTest.scala]:
```suggestion
      optionalResultExplanation = "Note that text indexes do not support multiple properties and that the index name must be unique.",
```

File: cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/SchemaIndexTest.scala [hunk 21]
```diff
@@ -206,6 +207,43 @@ class SchemaIndexTest extends DocumentingTestBase with QueryStatisticsTestSuppor
...
+      assertions = _ => assertIndexWithNameExists("node_index_name", "Person", List("surname"))
+    )
+    testQuery(
+      title = "Create a relationship text index",
+      text = "A named text index on a single property for all relationships that have a particular relationship type can be created with `CREATE TEXT INDEX index_name FOR ()-[r:TYPE]-() ON (r.property)`. " +
```
Comment (AlexicaWright) [inline cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/SchemaIndexTest.scala]:
```suggestion
      text = "A named text index on a single property for all relationships with a particular relationship type can be created with `CREATE TEXT INDEX index_name FOR ()-[r:TYPE]-() ON (r.property)`. " +
```

File: cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/SchemaIndexTest.scala [hunk 22]
```diff
@@ -206,6 +207,43 @@ class SchemaIndexTest extends DocumentingTestBase with QueryStatisticsTestSuppor
...
+        "Note that the index is not immediately available, but is created in the background.",
+      queryText = "CREATE TEXT INDEX rel_index_name FOR ()-[r:KNOWS]-() ON (r.interest)",
+      optionalResultExplanation = "Note that text indexes do not support multiple properties and that the index name needs to be unique.",
```
Comment (AlexicaWright) [inline cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/SchemaIndexTest.scala]:
```suggestion
      optionalResultExplanation = "Note that text indexes do not support multiple properties and that the index name must be unique.",
```

File: cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/SchemaIndexTest.scala [hunk 23]
```diff
@@ -206,6 +207,43 @@ class SchemaIndexTest extends DocumentingTestBase with QueryStatisticsTestSuppor
...
+      assertions = _ => assertIndexWithNameExists("rel_index_name", "KNOWS", List("interest"))
+    )
+    testQuery(
+      title = "Create a text index only if it does not already exist",
+      text = "If it is unknown if an index exists or not but we want to make sure it does, we add `IF NOT EXISTS`.",
```
Comment (AlexicaWright) [inline cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/SchemaIndexTest.scala]:
```suggestion
      text = "If it is not known whether an index exists or not but to ensure it does, add `IF NOT EXISTS`.",
```

**Review by AlexicaWright — COMMENTED** [review]
Looks great overall, but a few editorial comments.

File: cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/SchemaIndexTest.scala (hunk 20 above)
Comment (Hunterness) [reply cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/SchemaIndexTest.scala]: This I think is everywhere and not just here but I'll check

**Comment by Hunterness** [comment]
Several of your comments are not just where I have updated/added them now but all over the index (and maybe constraint) chapter, but I can see if I can update it everywhere

File: cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/SchemaIndexTest.scala (hunk 23 above)
Comment (Hunterness) [reply cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/SchemaIndexTest.scala]: Your sentence sounds odd to me, `... exists or not but to ensure it does ...` feels weird

Comment (Hunterness) [reply cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/SchemaIndexTest.scala]: I rewrote it someway lets see what you think

File: cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/ConstraintsTest.scala [hunk 24]
```diff
@@ -71,7 +71,7 @@ class ConstraintsTest extends DocumentingTestBase with SoftReset {
     )
     prepareAndTestQuery(
       title = "Create a unique constraint only if it does not already exist",
-      text = "If it is unknown if a constraint exists or not but we want to make sure it does, we add the `IF NOT EXISTS`. " +
+      text = "If it is not known whether a constraint exists or not, add `IF NOT EXISTS` to ensure it does. " +
```
Comment (AlexicaWright) [inline cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/ConstraintsTest.scala:74]: Yes, this works! In general, we try to steer away from using pronouns, like _we_ or _you_.

**Review by AlexicaWright — APPROVED** [review]
Great! Thank you @Hunterness !

**Review by AlexicaWright — APPROVED** [review]
All links work 👍

File: cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/SchemaIndexTest.scala [hunk 25]
```diff
@@ -248,7 +286,7 @@ class SchemaIndexTest extends DocumentingTestBase with QueryStatisticsTestSuppor
   @Test def drop_index_on_a_label_single_property() {
     prepareAndTestQuery(
       title = "Drop a single-property index",
-      text = "An index on all nodes that have a label and single property combination can be dropped with `DROP INDEX ON :Label(property)`.",
+      text = "An index on all nodes with a label and single property combination can be dropped with `DROP INDEX ON :Label(property)`.",
```
Comment (Hunterness) [inline cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/SchemaIndexTest.scala:289]: should probably specify that this is only valid for btree indexes

Comment (Hunterness) [reply cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/SchemaIndexTest.scala:289]: might be more updates around this that we want, might have done some of it as part of the range pr but might be good to see if any of that should be backported to this one

A bit dependent on the result of discussion with PM we have tomorrow around what we want to document

Comment (Hunterness) [reply cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/SchemaIndexTest.scala:289]: Will fix when adding range/point and not here

**Comment by Hunterness** [comment]
test failure related to https://github.com/neo4j/neo4j-documentation/pull/1266

**Comment by davidoliverSP2** [comment]
> test failure related to #1266

This has now been merged, so rebasing should resolve the build error 🤞
